onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -637,6 +644,56 @@ class PlannerImpl {
               if (found_in_outer_scope_location_map) {
                 plan_.SetLocation(static_cast<size_t>(index), iter->second);
               }
+            } else if (set_node_arg_has_explicit_consumer.count(index) == 0) {
+              // CASE 2: We see an implicit input without explicit consumers in the main graph,
+              // then set its location to be the device corresponding to the EP that the subgraph
+              // holding node has been partitioned to.
+
+              // The "ideal" solution is to set the location of its first "explicit" usage which may occur
+              // in any nested subgraph of the node, but that is potentially too costly to
+              // get at this stage (TODO: Investigate feasibility of this, see TODO in FinalizeSessionStateImpl() around this)
+
+              // Instead, we take a "less than ideal" route which is to set the location to be the device
+              // corresponding to the EP that the node is partitioned to. The hypothesis is that it is "most likely"
+              // that the implicit input will eventually be consumed on that device in a nested subgraph.
+
+              // The previous behavior was to default to CPU which will cause unnecessary copies when
+              // (1) The user invokes Run() with an OrtValue backed by non-CPU memory (eg CUDA) and
+              // the explicit consumer of the implicit input is on the non-CPU device in the subgraph
+              // (2) The user tries to IOBind implicitly consumed graph inputs (GH Issue 11254) and
```

Comment:
nit: 'explicit consumer of the implicit input' -> 'node in the subgraph that consumes the implicit input' might be easier to mentally unravel.

Reply:
Sure